subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/DependencyInsightReportTask.java

```diff
@@ -145,57 +153,71 @@ public abstract class DependencyInsightReportTask extends DefaultTask {
-                    );
-                }
-                configurationInternal.getResolutionStrategy().setIncludeAllSelectableVariantResults(true);
+                );
             }
-            configurationName = configuration.getName();
-            configurationDescription = configuration.toString();
-            zConfigurationAttributes = getProject().provider(configuration::getAttributes);
-
-            ProviderFactory providerFactory = getProject().getProviders();
-            ResolutionOutputsInternal resolutionOutputs = ((ResolvableDependenciesInternal) configuration.getIncoming()).getResolutionOutputs();
-            ResolutionResultProvider<VisitedGraphResults> graphResultsProvider = resolutionOutputs.getRawResults().map(ResolverResults::getVisitedGraph);
-            errorHandler.addErrorSource(providerFactory.provider(() ->
-                graphResultsProvider.getValue().getResolutionFailure()
-                    .map(Collections::singletonList)
-                    .orElse(Collections.emptyList()))
-            );
-            rootComponentProperty.set(providerFactory.provider(() -> {
-                // We do not use the public resolution result API to avoid throwing exceptions that we visit above
-                return graphResultsProvider.getValue().getResolutionResult().getRootSource().get();
-            }));
+            configurationInternal.getResolutionStrategy().setIncludeAllSelectableVariantResults(true);
         }
-        return rootComponentProperty;
+
+        ResolutionOutputsInternal resolutionOutputs = ((ResolvableDependenciesInternal) configuration.getIncoming()).getResolutionOutputs();
+        ResolutionResultProvider<VisitedGraphResults> graphResultsProvider = resolutionOutputs.getRawResults().map(ResolverResults::getVisitedGraph);
+        resolutionFailures = getProviderFactory().provider(() ->
+            graphResultsProvider.getValue().getResolutionFailure()
+                .map(Collections::singletonList)
+                .orElse(Collections.emptyList())
+        );
+        // We do not use the public resolution result API to avoid throwing exceptions that we visit above
+        return graphResultsProvider.getValue().getResolutionResult().getRootSource().get();
     }
 
     /**
-     * Selects the dependency (or dependencies if multiple matches found) to show the report for.
-     */
+     * The effective dependency spec selects the dependency (or dependencies if multiple matches found) to show the report for.
+     * The spec receives an instance of {@link DependencyResult} as parameter.
+     *
+     * @since 9.0
+     **/
     @Internal
-    @ToBeReplacedByLazyProperty(comment = "Should Spec<?> be lazy?")
-    public @Nullable Spec<DependencyResult> getDependencySpec() {
-        return dependencySpec;
+    @VisibleForTesting
+    @Incubating
+    protected Provider<Spec<DependencyResult>> getEffectiveDependencySpec() {
```

Comment:
❓ Do we need to expose that?

Reply:
I think yes. It's hard to even inspect the task without that. I can't find a feasible way of launching `DependencyInsightReportTaskSpec` without that being exposed